netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -117,14 +130,28 @@ public static FromChannelCredentialsResult from(ChannelCredentials creds) {
             new ByteArrayInputStream(tlsCreds.getPrivateKey()),
             tlsCreds.getPrivateKeyPassword());
       }
-      if (tlsCreds.getTrustManagers() != null) {
-        builder.trustManager(new FixedTrustManagerFactory(tlsCreds.getTrustManagers()));
-      } else if (tlsCreds.getRootCertificates() != null) {
-        builder.trustManager(new ByteArrayInputStream(tlsCreds.getRootCertificates()));
-      } // else use system default
+      Optional<TrustManager> x509ExtendedTrustManager;
       try {
-        return FromChannelCredentialsResult.negotiator(tlsClientFactory(builder.build()));
-      } catch (SSLException ex) {
+        if (tlsCreds.getTrustManagers() != null) {
+          builder.trustManager(new FixedTrustManagerFactory(tlsCreds.getTrustManagers()));
+          x509ExtendedTrustManager = tlsCreds.getTrustManagers().stream().filter(
+              trustManager -> trustManager instanceof X509ExtendedTrustManager).findFirst();
+        } else if (tlsCreds.getRootCertificates() != null) {
+          builder.trustManager(new ByteArrayInputStream(tlsCreds.getRootCertificates()));
+          x509ExtendedTrustManager = getX509ExtendedTrustManager(new ByteArrayInputStream(
+              tlsCreds.getRootCertificates()));
+        } else { // else use system default
+          TrustManagerFactory tmf = TrustManagerFactory.getInstance(
+              TrustManagerFactory.getDefaultAlgorithm());
+          tmf.init((KeyStore) null);
+          x509ExtendedTrustManager = Arrays.stream(tmf.getTrustManagers())
```

Comment:
Netty can be used on Android, although we discourage it. That means no java.util.streams.

That should apply to Optional as well... how is that already a dependency? Ah, it was added recently in #11600. We'll remove that.

Reply:
Done.